### ISSUE TITLE
"Lookup" ACL stopped working in 3.4

### ISSUE THREAD

**Issue by hdaeubler** [cyrusimap/cyrus-imapd#3528]
After a recent update to Cyrus imapd 3.4.1 the "List" commands fail to list folders with 'l' (lookup), but without 'r' (read) ACLs.
As I understand they should list them, as described in the "Access Control Lists Rights Reference", as well as in rfc4314.

This behaviour shows up only if "reverseacls" is enabled.

The responsible change is this:

```
--- cyrus-imapd-3.2.6/imap/mboxlist.c   2021-03-09 05:32:13.000000000 +0000
+++ cyrus-imapd-3.4.1/imap/mboxlist.c.orig      2021-05-10 06:34:24.000000000 +0000
@@ -693,11 +698,13 @@
     }
 }
 
-static int user_is_in(const strarray_t *aclbits, const char *user)
+static int user_can_read(const strarray_t *aclbits, const char *user)
 {
     int i;
     if (!aclbits) return 0;
     for (i = 0; i+1 < strarray_size(aclbits); i+=2) {
+        // skip ACLs without read bit
+        if (!strchr(strarray_nth(aclbits, i+1), 'r')) continue;
         if (!strcmp(strarray_nth(aclbits, i), user)) return 1;
     }
     return 0;
@@ -737,9 +744,10 @@
     if (oldusers) {
         for (i = 0; i+1 < strarray_size(oldusers); i+=2) {
             const char *acluser = strarray_nth(oldusers, i);
+            if (!strchr(strarray_nth(oldusers, i+1), 'r')) continue;
             if (!strcmpsafe(userid, acluser)) continue;
             if (strarray_find(admins, acluser, 0) >= 0) continue;
-            if (user_is_in(newusers, acluser)) continue;
+            if (user_can_read(newusers, acluser)) continue;
             mboxlist_racl_key(!!userid, acluser, name, &buf);
             r = cyrusdb_delete(mbdb, buf.s, buf.len, txn, /*force*/1);
             if (r) goto done;
@@ -750,9 +758,10 @@
     if (newusers) {
         for (i = 0; i+1 < strarray_size(newusers); i+=2) {
             const char *acluser = strarray_nth(newusers, i);
+            if (!strchr(strarray_nth(newusers, i+1), 'r')) continue;
             if (!strcmpsafe(userid, acluser)) continue;
             if (strarray_find(admins, acluser, 0) >= 0) continue;
-            if (user_is_in(oldusers, acluser)) continue;
+            if (user_can_read(oldusers, acluser)) continue;
             mboxlist_racl_key(!!userid, acluser, name, &buf);
             r = cyrusdb_store(mbdb, buf.s, buf.len, "", 0, txn);
             if (r) goto done;
```

**Comment by elliefm** [comment]
Hmm, am I misunderstanding the problem, or is there more to it than just reverseacls?  Here's a wip Cassandane test where I'm trying to reproduce the problem: https://github.com/cyrusimap/cassandane/compare/master...elliefm:v34/3528-list-lookup-only

The gist of it is like this:

```
[ as admin ]
C: 3 create shared/foo
S: 3 OK [MAILBOXID (876773a5-136b-4b19-8a1a-cb683f94d519)] Completed
C: 4 setacl shared/foo cassandane l
S: 4 OK Completed
```

```
[ as user cassandane ]
C: 3 list "" "*"
S: * LIST (\HasNoChildren) "/" INBOX
S: * LIST (\HasNoChildren) "/" "Shared Folders/shared/foo"
S: 3 OK Completed (0.000 secs 2 calls)
```

Which seems to be doing what you expect -- the shared/foo folder is returned in the LIST results, even though the only permission is "l" (no "r").

Very curiously, the cassandane user is also able to select the shared/foo folder, despite not having "r".

There's two tests here, one without reverseacls, one with, but both show the same behaviour.  I also tried with and without altnamespace but it didn't seem to make any difference.

When I run the test against the Cyrus master branch instead of 3.4, the "shared" parent folder is also listed, despite not having had an ACL set, so that's a whole other thing to think about.

**Comment by hdaeubler** [comment]
Yes it only happens with reverseacls enabled. Without reverseacls it works as expected.
It's a bit tricky to reproduce because you have to make sure the RACL records in mailboxes.db are really rebuilt when changing the reverseacl setting. I did this by

```
[ set reverseacls: 0 ]
ctl_cyrusdb -r

[ set reverseacls: 1 ]
ctl_cyrusdb -r
```
I also checked with
```
cyr_dbtool /mail/var/imap/mailboxes.db skiplist show
```
to see if the RACLs are present depending on whether the permission is 'l' or 'lr'. It was only there with permission 'lr'.

**Comment by elliefm** [comment]
Cassandane builds a new environment for each test, so there's no danger of the old reverseacl state still hanging around or not between tests.

But my tests are on a shared mailbox, because that seemed the most likely use case for an "l but not r" access control, but I wonder if that was a bad assumption...

Is this happening for the user's own mailboxes (that they only have 'l' rights on, for some reason)?  Is it happening for some-other-user's mailboxes?  Shared mailboxes?

**Comment by hdaeubler** [comment]
Hmm, ok, it's strange that the tests fail to see the problem.

It happens for other user's mailboxes as well as shared mailboxes, as soon as they have 'l' but not 'r' and reverseacls is on and it definitely was not a problem in 2.6.* releases. I first saw it in 3.4.*

I just reproduced it "by hand" to make sure once again it is really there.

**Comment by elliefm** [comment]
Interesting, I'll try adding tests for other user's mailboxes and see if I can reproduce it like that.  Though I wonder if there's some other setting that's impacting this.  What are your settings for altnamespace, unixhierarchysep, virtdomains...?

Are you able to contrive a sample reproduction session that you can paste here without exposing any real account or mailbox names?  Like the ones I pasted a few comments back.

**Comment by elliefm** [comment]
Oh also, what's your setting for improved_mboxlist_sort?

(I don't have any particular insight as to which settings might be changing the behaviour here, but altnamespace/unixhierarchysep/virtdomains/improved_mboxlist_sort are "the usual suspects", so that's as good a place to start as any)

**Comment by elliefm** [comment]
And another, what's your setting for "reverseuniqueids"?

**Comment by elliefm** [comment]
Oh, hey, I've reproduced the problem when using another user's mailbox (as opposed to a global shared mailbox).

Without reverseacl:

```
[ as admin, "user/other" inbox already exists]
C: 3 create user/other/foo
S: 3 OK [MAILBOXID (31028b03-374a-4583-9dbe-faaab0efd87d)] Completed
C: 4 setacl user/other/foo cassandane l
S: 4 OK Completed
```

```
[as user cassandane]
C: 3 list "" "*"
S: * LIST (\HasNoChildren) "/" INBOX
S: * LIST (\HasNoChildren) "/" "Other Users/other/foo"
S: 3 OK Completed (0.000 secs 2 calls)
```

With reverseacl:

```
[as admin, "user/other" inbox already exists]
C: 3 create user/other/foo
S: 3 OK [MAILBOXID (a874da47-d14a-491a-833a-3d9b82bb27ba)] Completed
C: 4 setacl user/other/foo cassandane l
S: 4 OK Completed
```

```
[as user cassandane]
C: 3 list "" "*"
S: * LIST (\HasNoChildren) "/" INBOX
S: 3 OK Completed (0.000 secs 1 calls)
```

And adding the "r" permission in the reverseacl:on case allowed the mailbox to be listed, as reported.

**Comment by elliefm** [comment]
Now that I can reproduce it, bisect finds it was broken by https://github.com/cyrusimap/cyrus-imapd/commit/ff1442271ae144fb84fcf58f0e258f54d8c11164

Which makes some sense -- I guess when reverseacl is on, LIST uses it to optimise lookups, but since this commit, it's only in reverseacl if there's an "r" permission.

**Comment by hdaeubler** [comment]
Thanks for your effort to check this out.
Regards

**Comment by elliefm** [comment]
This is now fixed on the 3.4 and master branches, and will be fixed in future releases.

In the meantime, you can get the fix early by applying this patch: https://github.com/cyrusimap/cyrus-imapd/commit/b5d63ae71409a20d3b9a35abf96ca91031048d87.patch
